src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -39,6 +39,24 @@ public static string GetUnescapedString(ReadOnlySpan<byte> utf8Source, int idx)
             return utf8String;
         }
 
+        // TODO: Similar to escaping, replace the unescaping logic with publicly shipping APIs from https://github.com/dotnet/corefx/issues/33509
+        public static byte[] GetUnescapedBytes(ReadOnlySpan<byte> utf8Source, int idx)
+        {
+            byte[] unescapedArray = null;
+
+            Span<byte> utf8Unescaped = utf8Source.Length <= JsonConstants.StackallocThreshold ?
+                stackalloc byte[utf8Source.Length] :
+                (unescapedArray = ArrayPool<byte>.Shared.Rent(utf8Source.Length));
+
+            Unescape(utf8Source, utf8Unescaped, idx, out int written);
+            Debug.Assert(written > 0);
+
+            utf8Unescaped = utf8Unescaped.Slice(0, written);
+            Debug.Assert(!utf8Unescaped.IsEmpty);
+
+            return utf8Unescaped.ToArray();
```

Comment:
This call to `ToArray()` undoes the allocation savings we tried to do above using stackalloc/ArrayPool.

We could get away with it in the string case since transcoding allocates the string anyway. Here, we probably want to figure out how we can interleave this code with the call to the parser to get the .NET type (whether it's `Guid` or `DateTime`).